### PR TITLE
Add Reshape to InsertQuantizeOnDtypeMismatch

### PR DESCRIPTION
**Pull request by jinevening**
This adds Reshape Op to InsertQuantizeOnDtypeMismatch.

TICO-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

### CODE REVIEW

**Comment by mhs4670go** [comment]
Style-check failed.

**Comment by jinevening** [comment]
Updated the PR. I made https://github.com/Samsung/TICO/issues/36 for automatic formatting.

**Review by mhs4670go — APPROVED** [review]
LGTM

File: tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py [hunk 1]
```diff
@@ -271,6 +272,39 @@ def _insert_quantize_op_after(node):
...
+
+                if QPARAM_KEY not in node.meta:
+                    continue
+
+                if qparam_dtype(inp) == qparam_dtype(node):
+                    continue
+
+                if qparam_dtype(inp) == "int16" and qparam_dtype(node) == "uint8":
+                    # A new Quantize Op (s16 to u8) is inserted before (not after)
+                    # reshape Op to reduce tensor size ealier
```
Comment (dayo09) [inline tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py:297]: I am new to this parts of codes, so could you please explain why int16->uint8 requires `insert quantize op before` and uint8->int16 requires `insert quantize op after` functions?

File: tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py [hunk 2]
```diff
@@ -271,6 +272,39 @@ def _insert_quantize_op_after(node):
...
+                    quantize = _insert_quantize_op_before(node, inp)
+
+                    quantize.meta[QPARAM_KEY] = copy.deepcopy(node.meta[QPARAM_KEY])
+                    node.meta[QPARAM_KEY] = _u8_to_i16(node.meta[QPARAM_KEY])
```
Comment (miusic) [inline tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py:301]: May I ask some more details for the reason why the dtype of node was changed to `s16`?

IIUC, isn't it enough to insert `quantize` and not to change the dtype of the node

https://github.com/Samsung/TICO/blob/f0aeefcd34ae306392f37e8525070c792f8dddd2/tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py#L99-L104

Comment (miusic) [reply tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py:301]: Based on my understanding, below is the graph before this pass and after this pass

### Before
```
Input(S16) - Reshape(U8)
```

### After
```
Input(S16) - Quantize(U8) - Reshape(S16)
```

File: tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py (hunk 1 above)
Comment (jinevening) [reply tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py:297]: Let me explain the rationale. For int16->uint8,

1. Insert Quantize before Reshape.

```
Predecessor (int16)-> Quantize (uint8) -> Reshape (uint8) -> ...
```

2. Insert Quantize after Reshape.

```
Predecessor (int16)-> Reshape (int16) -> Quantize (uint8) -> ...
```

Comparing 1) and 2), the difference is that Reshape operation is conducted in uint8 or int16.  I decided to go with 1), which does Reshape in uint8, for faster execution of Reshape. Note that Reshape Op does not change the value, so its dytpe does not affect accuracy.

uint8->int16 case is the opposite of the above example.

File: tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py (hunk 2 above)
Comment (jinevening) [reply tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py]: Nice catch! I've updated the code not to update Reshape's qparam. Permute was working incorrectly, so I updated it too.

File: test/quantization/pass/test_insert_quantize_on_dtype_mismatch.py [hunk 3]
```diff
@@ -107,38 +117,92 @@ def setup(self, mod: torch.nn.Module, target_op, scale=1.0, zp=0, dtype="uint8")
 
     def run_test(self):
         # Before pass
-        self.assertNotEqual(self.target.meta[QPARAM_KEY].dtype, self.dtype)
+        # InsertQuantizeOnDtypeMismatch works for the case when input dtype does not match with output dtype
```
Comment (dayo09) [inline test/quantization/pass/test_insert_quantize_on_dtype_mismatch.py:120]:
```suggestion
        # InsertQuantizeOnDtypeMismatchTest works for the case when input dtype does not match with output dtype
```

(optional) The pass `InsertQuantizeOnDtypeMismatch` works on all cases, it's `InsertQuantizeOnDtypeMismatchTest` that works for mismatch case. So to be the comment technically right.

File: tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py (hunk 1 above)
Comment (dayo09) [reply tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py:296]: @jinevening Thank you for the clear explanation :-D!!

**Review by dayo09 — DISMISSED** [review]
LGTM

**Review by miusic — DISMISSED** [review]
LGTM! 👍

File: test/quantization/pass/test_insert_quantize_on_dtype_mismatch.py [hunk 4]
```diff
@@ -107,38 +117,92 @@ def setup(self, mod: torch.nn.Module, target_op, scale=1.0, zp=0, dtype="uint8")
...
 
 class PermuteTest(InsertQuantizeOnDtypeMismatchTest):
     def test_i8o16(self):
-        self.setup(SimplePermute(), torch.ops.aten.permute.default, dtype="uint8")
+        self.setup(
+            SimplePermute(),
+            torch.ops.aten.permute.default,
+            input_dtype="uint8",
+            desired_dtype="uint8",
+        )
```
Comment (seockho-kim) [inline test/quantization/pass/test_insert_quantize_on_dtype_mismatch.py:145]: I have a question.
What the meaning of test_i8o16? input 8bit, output 16bit.
Then, desired_dtype "uint8" looks a little weird to me.

Comment (dayo09) [reply test/quantization/pass/test_insert_quantize_on_dtype_mismatch.py:145]: IIUC, it's `precedence(uint8) - permute(int16)` case. Thus, it must be converted as `precedence(uint8) - permute(uint8, which is changed from int16) - quantize(int16 -> uint8, thus its qparam is int16) `.

desired_dtype means its target node's desired dtype. In this example, target node is permute.
Therefore desired_dtype is uint8.

Comment (jinevening) [reply test/quantization/pass/test_insert_quantize_on_dtype_mismatch.py:145]: @dayo09 Thanks for explanation :) It seems that the code is a bit difficult to understand. I added comments in the pass code

```python
    - NOTE For some cases, Quantize Op is inserted before the operators.

    Let's assume Reshape Op's input is int16 and output is uint8. There are two possible places to insert
    Quantize Op.

    1. Insert Quantize before Reshape.

    ```
    Predecessor (int16)-> Quantize (uint8) -> Reshape (uint8) -> ...
    ```

    2. Insert Quantize after Reshape.

    ```
    Predecessor (int16)-> Reshape (int16) -> Quantize (uint8) -> ...
    ```

    Comparing 1) and 2), the difference is that Reshape operation is conducted in uint8 or int16.
    We go with 1), which does Reshape in uint8, for faster execution. Note that Reshape Op does not
    change the value, so its dytpe does not affect accuracy.
```

and in the test code.

```python
    # Quantize Op is inserted before Permute, so desired_dtype of Reshape is uint8
    # After conversion)
    # Input (s16) -> Quantize (u8) -> Permute (u8)
    def test_i16o8(self):
        self.setup(
            SimplePermute(),
            torch.ops.aten.permute.default,
            input_dtype="int16",
            desired_dtype="uint8",
        )
        self.run_test()
```

**Review by dayo09 — APPROVED** [review]
LGTM

File: tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py [hunk 5]
```diff
@@ -102,6 +103,26 @@ class InsertQuantizeOnDtypeMismatch(PassBase):
...
+
+    2. Insert Quantize after Reshape.
+
+    ```
+    Predecessor (int16)-> Reshape (int16) -> Quantize (uint8) -> ...
+    ```
+
+    Comparing 1) and 2), the difference is that Reshape operation is conducted in uint8 or int16.
+    We go with 1), which does Reshape in uint8, for faster execution. Note that Reshape Op does not
+    change the value, so its dytpe does not affect accuracy.
```
Comment (dayo09) [inline tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py:125]:
```suggestion
    change the value, so its dtype does not affect accuracy.
```

**Review by miusic — APPROVED** [review]

**Review by mhs4670go — APPROVED** [review]
LGTM
